File: templates/python/dca.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""决策曲线分析 DCA（净获益）—— 卡片 06_预测模型评价图/决策曲线_DCA.md
比较模型 vs treat-all vs treat-none 的 net benefit。纯计算，无特殊依赖。
合成预测，可独立运行。改 # >>> PARAM 区换 y_true/y_prob。"""
import os, sys, warnings; warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np, matplotlib.pyplot as plt
from _common import set_theme, save_fig, pal
set_theme(); np.random.seed(7)

# >>> PARAM ------------------------------------------------------
NAME = "tpl_dca"
TITLE = "Decision curve analysis"; XLAB = "Threshold probability"; YLAB = "Net benefit"
MODEL_COLOR = pal("med_case_control")["low"]
TMIN, TMAX = 0.01, 0.6
FIGSIZE = (3.8, 3.2)
# 换真实数据： y_true = df["y"].values; y_prob = model.predict_proba(X)[:,1]
n = 2000
y_prob = np.random.beta(2, 4, n)
y_true = np.random.binomial(1, y_prob)
# <<< PARAM ------------------------------------------------------

def net_benefit(y, p, t):
    tp = ((p >= t) & (y == 1)).sum(); fp = ((p >= t) & (y == 0)).sum(); n = len(y)
    return tp / n - fp / n * (t / (1 - t))

thr = np.linspace(TMIN, TMAX, 100); prev = y_true.mean()
nb_model = [net_benefit(y_true, y_prob, t) for t in thr]
nb_all = [prev - (1 - prev) * (t / (1 - t)) for t in thr]
fig, ax = plt.subplots(figsize=FIGSIZE)
ax.plot(thr, nb_model, color=MODEL_COLOR, lw=1.8, label="Model")
ax.plot(thr, nb_all, color="#999999", lw=1, ls="-", label="Treat all")
ax.axhline(0, color="black", lw=1, ls="--", label="Treat none")
ax.set_ylim(min(-0.02, min(nb_model)), max(nb_model) * 1.15 + 1e-3)
ax.set_title(TITLE); ax.set_xlabel(XLAB); ax.set_ylabel(YLAB)
ax.grid(False); ax.legend(fontsize=8)
save_fig(fig, NAME)

File: templates/r/dca.R
## 决策曲线分析 DCA（净获益）—— 卡片 06_预测模型评价图/决策曲线_DCA.md
## 比较 模型 vs treat-all vs treat-none。纯计算。合成预测，可独立运行： Rscript templates/r/dca.R
.args <- commandArgs(trailingOnly = FALSE)
.sd <- dirname(normalizePath(sub("^--file=", "", .args[grep("^--file=", .args)])))
source(file.path(.sd, "_common.R"))
set.seed(2026)

# >>> PARAM ------------------------------------------------------
NAME <- "tpl_dca"; TITLE <- "Decision curve analysis"
XLAB <- "Threshold probability"; YLAB <- "Net benefit"
MODEL_COLOR <- unname(pal("med_case_control")["low"])
TMIN <- 0.01; TMAX <- 0.6
# 换真实数据： y <- df$y; prob <- predict(model, type = "response")
n <- 2000; prob <- rbeta(n, 2, 4); y <- rbinom(n, 1, prob)
# <<< PARAM ------------------------------------------------------

nb <- function(t){ tp <- sum(prob >= t & y == 1); fp <- sum(prob >= t & y == 0)
  tp / n - fp / n * (t / (1 - t)) }
thr <- seq(TMIN, TMAX, length.out = 100); prev <- mean(y)
d <- rbind(
  data.frame(thr = thr, nb = sapply(thr, nb),                      strat = "Model"),
  data.frame(thr = thr, nb = prev - (1 - prev) * (thr / (1 - thr)), strat = "Treat all"),
  data.frame(thr = thr, nb = 0,                                     strat = "Treat none"))
d$strat <- factor(d$strat, levels = c("Model", "Treat all", "Treat none"))
p <- ggplot(d, aes(thr, nb, color = strat, linetype = strat)) +
  geom_line(linewidth = .9) +
  scale_color_manual(values = c(Model = MODEL_COLOR, `Treat all` = "#999999", `Treat none` = "black")) +
  scale_linetype_manual(values = c(Model = 1, `Treat all` = 1, `Treat none` = 2)) +
  coord_cartesian(ylim = c(min(-0.02, min(sapply(thr, nb))), max(sapply(thr, nb)) * 1.15 + 1e-3)) +
  labs(x = XLAB, y = YLAB, color = NULL, linetype = NULL, title = TITLE)
save_gg(p, NAME, w = 100, h = 82)
